### PR TITLE
Standardize OpenType Features across faces

### PR DESCRIPTION
**Pull request by skef**
This request involves a large degree of change and will probably take a while to work through the reviewing. So I'm going to document the technical part now and the design part a bit later. 

The technical part relates to the GSUB tables. With these changes, the features of the main font faces (Serif-Regular, Serif-Italic, Serif-Bold, Serif-BoldItalic, Serif-Semibold, Serif-SemiboldItalic,, Sans-Regular, Sans-Italic, Sans-Bold, SerifDisplay-Regular) are now standardized on the model of Serif-Regular. The details and implementation are in sources/features. There are now feature files for all tables other than the experimental combining accents for superiors in Serif-Regular, and the weird Elvin variant in Sans-Bold attached to ss05 (the latter of which I recommend be put "on ice" in a future change). The `README.md` in that directory has more details. 

However, there is something important that document does not mention: `reload_features.py` will not run correctly without some fixes to FontForge, mainly those in fontforge/fontforge#3344 and fontforge/fontforge#3350 . Both of those are merged, so code compiled from the current master will work fine. And anyway, the current `.sfd` files match the contents of the feature file tables in `sources/features`, so it is not necessary to run the script other than to test it. 

The design part of the review involves assessing the glyphs added to allow the feature set to converge across the faces. Well, almost converge: Sans and Serif have somewhat different needs, and Display-Regular does not yet have smallcaps. Other than that, the substitutions are now consistent. 

For your convenience, here is a zip with the output of `sfddiff` comparing the normalized files of this push with those of the current master:  [sfddiff_out.zip](https://github.com/libertinus-fonts/libertinus/files/2720625/sfddiff_out.zip)

Note that there are many glyphs marked as missing in Sans-Regular and Sans-Bold. This was due to a re-encoding snafu that removed a number of "glyphs" that had no outlines or references but only color entries. I can stitch those back in if really necessary, but don't want to bother unless it matters. 

Other than those, here are the explanations for the other *missing* glyphs:

1. Sans-*: zero.slashfitted is now zero.slash.fitted, Q_u is now Q.u (and implemented differently), K.alt and R.alt were swashes copied from the serif fonts
2. Sans-Italic: uniE04F is now t_z (and updated)
3. Serif-BoldItalic: The .superiors were moved into the 1DXX unicode slots. scommaaccent.sc was renamed, uniE02F was renamed W.alt
4. Serif-Regular: uni01E2.sc was renamed
5. Serif-Italic: The .superiors were moved into the 1DXX unicode slots.
6. Serif-SemiboldItalic: uniE04D was renamed
7. Serif-Semibold: The .superiors ..., tcedilla.sc was renamed.

Beyond this, some glyphs have been changed and many have been added. I'll try to provide a better explanation of the changes (or at least an easier route to reviewing them) soon. For now, the main thing to know is that the superiors and inferiors added to Serif-SemiboldItalic are meant mostly as placeholders. The situation with those glyphs outside of Serif-Regular is a mess anyway, and needs an overhaul. Those added or adapted are about as close as those in other faces, such as Serif-BoldItalic. 

If and when this is merged it will resolve or almost resolve the following bugs:

#191 , #47 , #29 

With #29 the glyphs are now present and substituting as in Serif-Italic, but we should nail down the language/script combinations and determine if we need an alternate for uni0453. 

I suggest starting with a look from the high level, and trying some things out, and then clarifying what more information or changes you would like to see before this can go forward.

### CODE REVIEW

**Comment by skef** [comment]
Oh, the `case` lookup has been reorganized as I mentioned and you (broadly) agreed with in #236

**Comment by skef** [comment]
I Corrected the map from /degree /r to /degree /F in liga_sans.fig and rebased. If and when I find issues I'll continue to correct, rebase, force-push to my branch, and not the change here unless you advise some other process.

**Comment by skef** [comment]
I fixed a few missing references in Serif-BoldItalic, resynced with the master, and repushed. 

Here is the promised glyph review helper, in the form of a zip file containing an `index.html` to read and individual SVGs of the modified glyphs: [glyph_review.zip](https://github.com/libertinus-fonts/libertinus/files/2729454/glyph_review.zip)

**Comment by skef** [comment]
Any feedback on this in terms of what it would take to get some of the changes integrated? Would it be better to break the content down in a different way?

**Comment by khaledhosny** [comment]
Sorry for not responding earlier. But the changes are big and I need time studying and verifying them.

One aspect I don’t like (without having checked the changes) is having to separately run a script to apply the feature files to the SFD; I’d like to have a single source of truth and features files would be my preference for GSUB features, so if we move stuff there it should be gone from SFD files and the features files are applied at build time.

I understand this requires and new, unreleased FontForge, so that is a big limitation. One alternative would to switch to FontTools’ feaLib for compiling the feature files (like I do in my Amiri font), but this is likely to be tricky with the SFD mix and requires more thought and work.

**Comment by skef** [comment]
A few thoughts on the "authority" issue, none conclusive:

1. In my mind the most relevant observation is that the distance (in time and effort) from the "draft" in this request and any of the end-result you describe is small. Because feature files support `include` and tables defined outside of lookups, one could right now define "top-level" feature files for each of the fonts quite easily. And for information-sharing purposes having the tables that differ in separate files like they are now make sense. So the issues are important but the differences in implementation are small. 

2. I agree that the feature files should be the "authority" going forward. The question is how to achieve that. 

3. At least at this point I would argue against stripping the GSUB content from the files. I know the Metrics window isn't everyone's cup of tea but I've found it quite useful for testing how different combination of features work together. I feel that taking those capabilities away from potential contributors would be a net negative. 

4. If that's right, the problem becomes how to make the feature files authoritative while still retaining the GSUB content. But this is easy and "make check" already provides a model: a hook can be added to the Makefile to reload all the tables (however) and report any differences as errors. In fact the only thing that argues against adding this to `make check` itself is that for a while it will require a very recent version of fontforge to run, so it's best not to require it on the "developer side", while making it available as an option. 

5. If something like 4 is the way to go, then I am personally agnostic about a choice between my script and using "top level" feature files. It's just as easy to strip out all GSUB content and then merge a whole file (as far as I know -- I haven't tried). It would have the advantage of making the files also authoritative for scripts and languages. The disadvantage would be limitation in the table names, but that doesn't seem like a significant problem. I'm therefore happy to rewrite things to work that way if that's your preference.  

tl;dr: I favor making the feature files authoritative while keeping the GSUB content on the model of "make check".

**Comment by skef** [comment]
Oh, I'll add: I wrote the code in #249 because I anticipated something like this model. If the GSUB content remains but the feature files are routinely reloaded, it helps to have a canonical ordering of feature content to limit the diff to substantial changes.

**Comment by khaledhosny** [comment]
Closing in favor of #258.
